runtime/src/stakes/serde_stakes.rs

```diff
@@ -2,13 +2,74 @@ use {
     super::{StakeAccount, Stakes, StakesEnum},
     crate::stake_history::StakeHistory,
     im::HashMap as ImHashMap,
-    serde::{ser::SerializeMap, Serialize, Serializer},
+    serde::{ser::SerializeMap, Deserialize, Deserializer, Serialize, Serializer},
     solana_sdk::{clock::Epoch, pubkey::Pubkey, stake::state::Delegation},
     solana_stake_program::stake_state::Stake,
     solana_vote::vote_account::VoteAccounts,
     std::sync::Arc,
 };
 
+/// Wrapper struct with custom serialization to support serializing
+/// `Stakes<StakeAccount>` as `Stakes<Stake>` without doing a full deep clone of
+/// the stake data. Serialization works by building a `Stakes<&Stake>` map which
+/// borrows `&Stake` from `StakeAccount` entries in `Stakes<StakeAccount>`. Note
+/// that `Stakes<&Stake>` still copies `Pubkey` keys so the `Stakes<&Stake>`
+/// data structure still allocates a fair amount of memory but the memory only
+/// remains allocated during serialization.
+#[cfg_attr(feature = "frozen-abi", derive(AbiExample, AbiEnumVisitor))]
+#[derive(Debug, Clone)]
+pub enum SerdeStakesToStakeFormat {
+    Stake(Stakes<Stake>),
+    Account(Stakes<StakeAccount>),
+}
+
+#[cfg(feature = "dev-context-only-utils")]
+impl PartialEq<Self> for SerdeStakesToStakeFormat {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Stake(stakes), Self::Stake(other)) => stakes == other,
+            (Self::Account(stakes), Self::Account(other)) => stakes == other,
+            (Self::Stake(stakes), Self::Account(other)) => {
+                stakes == &Stakes::<Stake>::from(other.clone())
+            }
+            (Self::Account(stakes), Self::Stake(other)) => {
+                other == &Stakes::<Stake>::from(stakes.clone())
+            }
+        }
+    }
+}
+
+impl From<SerdeStakesToStakeFormat> for StakesEnum {
+    fn from(stakes: SerdeStakesToStakeFormat) -> Self {
+        match stakes {
+            SerdeStakesToStakeFormat::Stake(stakes) => Self::Stakes(stakes),
+            SerdeStakesToStakeFormat::Account(stakes) => Self::Accounts(stakes),
+        }
+    }
+}
+
+impl Serialize for SerdeStakesToStakeFormat {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            Self::Stake(stakes) => stakes.serialize(serializer),
+            Self::Account(stakes) => serialize_stake_accounts_to_stake_format(stakes, serializer),
```

Comment:
Now we are doing a memory efficient serialization approach which doesn't need to allocate a new stakes map before serialization